File: 2020/13/solution.exs
IO.write("\e[H\e[2J")
IO.puts("Running #{__ENV__.file}...")

lines = __DIR__ <> "/" <> "input.txt"
  |> File.read!()
  |> String.split("\n", trim: true)

[timestamp, busses] = lines
timestamp = String.to_integer(timestamp)
part_one_busses = String.split(busses, ",", trim: true)
  |> Enum.filter(fn s -> s != "x" end)
  |> Enum.map(&String.to_integer/1)

part_two_busses = String.split(busses, ",", trim: true)
  |> Enum.with_index
  |> Enum.filter(fn {v, _} -> v != "x" end)
  |> Enum.map(fn {v, i} -> {String.to_integer(v), i} end)

defmodule PartOne do
  def run(timestamp, busses) do
    times = Enum.map(busses, fn bus ->
      if rem(timestamp, bus) == 0 do
        timestamp
      else
        { bus, bus * (div(timestamp, bus) + 1) }
      end
    end)
    {id, time} = Enum.min_by(times, fn {_, time} -> time end)
    difference = time - timestamp
    difference * id
  end
end

defmodule PartTwo do
  def run(busses, product \\ 0, total \\ 0, coefficient \\ 0)
  def run([], _, total, _), do: total
  def run([{bus, _} | tail], 0, 0, 0), do: run(tail, bus, bus, 0)
  def run([{b, i} | tail], p, t, c) when rem(t + p * c + i, b) == 0, do: run(tail, p * b, t + p * c, 0)
  def run([{bus, index} | tail], product, total, coefficient) do
    run([{bus, index} | tail], product, total, coefficient + 1)
  end
end


PartOne.run(timestamp, part_one_busses) |> IO.inspect
PartTwo.run(part_two_busses) |> IO.inspect

File: 2020/13/solution.go
package main

import (
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
)

type busInfo struct {
	bus   int
	index int
}

func readInput() (int, []int, []busInfo) {
	input, err := ioutil.ReadFile("2020/13/input.txt")
	if err != nil {
		panic(err)
	}
	lines := strings.Split(string(input), "\n")
	timestamp, _ := strconv.Atoi(lines[0])
	busses := strings.Split(lines[1], ",")

	var partOneBusses []int
	var partTwoBusses []busInfo
	for i, bus := range busses {
		if bus != "x" {
			number, _ := strconv.Atoi(bus)
			partOneBusses = append(partOneBusses, number)
			partTwoBusses = append(partTwoBusses, busInfo{number, i})
		}
	}

	return timestamp, partOneBusses, partTwoBusses
}

func partOne(timestamp int, busses []int) int {
	bus := busses[0]
	min := timestamp * 2
	for _, number := range busses {
		times := timestamp / number
		arrival := (times + 1) * number
		if arrival < min {
			bus = number
			min = arrival
		}
	}
	return (min - timestamp) * bus
}

func remainder(a int, b int) int {
	return a - ((a / b) * b)
}

func partTwo(busses []busInfo) int {
	total := 0
	product := 1
	for _, bi := range busses {
		co := 0
		for remainder(total+(co*product)+bi.index, bi.bus) > 0 {
			co++
		}
		total = total + (co * product)
		product = product * bi.bus
	}
	return total
}

func main() {
	timestamp, partOneBusses, partTwoBusses := readInput()

	fmt.Println(partOne(timestamp, partOneBusses))
	fmt.Println(partTwo(partTwoBusses))
}
